File: 3831-find-x-value-of-array-i/find-x-value-of-array-i.py
from typing import List

class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)

        result = [0] * k
        prevCount = [0] * k

        for i in range(n):
            currCount = [0] * k

            currElement = nums[i] % k

            currCount[currElement] += 1

            for oldRem in range(k):
                newRem = (oldRem * nums[i]) % k

                currCount[newRem] += prevCount[oldRem]

            prevCount = currCount

            for rem in range(k):
                result[rem] += prevCount[rem]

        return result

File: 3831-find-x-value-of-array-i/find-x-value-of-array-i.java
class Solution {
    public long[] resultArray(int[] nums, int k) {
        int n = nums.length;

        long[] result = new long[k];
        long[] prevCount = new long[k];

        for (int i = 0; i < n; i++) {

            long[] currCount = new long[k];

            int currElement = nums[i] % k;

            currCount[currElement]++;

            for (int oldRem = 0; oldRem < k; oldRem++) {

                int newRem = (int)(((long) oldRem * nums[i]) % k);

                currCount[newRem] += prevCount[oldRem];
            }

            prevCount = currCount;

            for (int rem = 0; rem < k; rem++) {
                result[rem] += prevCount[rem];
            }
        }

        return result;
    }
}